<a href="https://colab.research.google.com/github/SarahLares/introduction-to-deep-learning-with-keras-Coursera/blob/master/build_a_regression_model_in_keras_Part_C.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Regresioon Model in Keras

<div class="alert alert-block alert-info" style="margin-top: 20px">

<font size = 3>

In this course project, a regression model was constructed using the Keras deep learning library, and then experimented with increasing the number of training epochs and changing the number of hidden layers, and discussing how changing these parameters affects model performance. 

</font>
</div>

In [0]:
import pandas as pd
import numpy as np

Importing the data as a dataframe

In [0]:
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  ...  Fine Aggregate  Age  Strength
0   540.0                 0.0      0.0  ...           676.0   28     79.99
1   540.0                 0.0      0.0  ...           676.0   28     61.89
2   332.5               142.5      0.0  ...           594.0  270     40.27
3   332.5               142.5      0.0  ...           594.0  365     41.05
4   198.6               132.4      0.0  ...           825.5  360     44.30

[5 rows x 9 columns]

So the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa. 

Split data into predictors and target. The target variable in this problem is the concrete sample strength. Therefore, our predictors will be all the other columns.

In [0]:
concrete_data_columns = concrete_data.columns

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

## C. Increate the number of epochs (5 marks)



<div class="alert alert-block alert-info" style="margin-top: 20px">

<font size = 3>

Repeat Part B but use 100 epochs this time for training.

</font>
</div>

In [0]:
#Normalization of the predictors
predictors_norm = (predictors - predictors.mean()) / predictors.std()

In [0]:
n_cols = predictors_norm.shape[1]

In [0]:
import keras
from keras.models import Sequential
from keras.layers import Dense

Using TensorFlow backend.


In [0]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

### 1. Randomly split the data into a training and test sets by holding 30% of the data for testing. 

You can use the train_test_splithelper function from Scikit-learn.


In [0]:
from sklearn.model_selection import train_test_split

In [0]:
X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.30, random_state=42)

### 2. Train the model on the training data using 100 epochs.

In [0]:
model = regression_model()

In [0]:
model.fit(X_train, y_train, epochs=100, verbose=0)

In [0]:
y_pred = model.predict(X_test)

### 3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. 
You can use the mean_squared_error function from Scikit-learn.

In [0]:
from sklearn.metrics import mean_squared_error

In [0]:
mse = mean_squared_error(y_test,y_pred)
mse

151.10708610404603

### 4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.

In [0]:
MSE = []

In [0]:
for i in range(50):
  X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.30, random_state=42)
  model.fit(X_train, y_train, epochs=50, verbose=0)
  y_pred = model.predict(X_test)
  mse = mean_squared_error(y_test,y_pred)
  MSE.append(mse)

### 5. Report the mean and the standard deviation of the mean squared errors.

In [0]:
print('The mean of the 50 mean squared error is: {:.3f}'.format(np.mean(MSE)))
print('The standar deviation of the 50 mean squared error is: {:.3f}'.format(np.std(MSE)))

The mean of the 50 mean squared error is: 5770.381
The standar deviation of the 50 mean squared error is: 25798.548


**Compared with the values ​​obtained in part B, where the predictors was normalized and the number of epoch are 50, we can see that with the same data and change the number the epoch to 100 the values ​​are greater than those obtained in part B.**